### DISCUSSION TITLE
Contact Shadow Flickering

### DISCUSSION THREAD

**Discussion by Nolram12345** [google/filament#7691] [General]
Hello!
We've noticed during that development that there is a frequent issue with contact shadow flickering on sloped surfaces.

Here's the issue visualised in a test scene in the GLTF viewer:
https://github.com/google/filament/assets/37701550/ad080d02-4ffa-44e2-8caa-d4811a64aff8

1. Is this a known issue?
2. Is there a workaround?

To me personally this seems like something that could be addressed with a slope bias for contact shadows.

**Comment by Nolram12345** [comment]
Update: Further experimentation also reveals that the issue gets **significantly** worse with higher sample counts for the contact shadows.

https://github.com/google/filament/assets/37701550/f29be451-3ed5-496f-aae0-833481dc0bf4

**Reply by romainguy** [reply]
On what OS with what backend does this happen? I am not able to reproduce it.

**Reply by Nolram12345** [reply]
Vulkan & OpenGL. Windows 10 22H2.
Reproduced on both AMD and NVIDIA GPUs.

**Reply by Nolram12345** [reply]
I'm currently working on a lightweight scene that can reproduce the issue for further testing, as my above test scene contains licensed assets.

**Reply by Nolram12345** [reply]
Here's a GLB and the settings used to reproduce the issue.
[contact_shadow_issue.zip](https://github.com/google/filament/files/14733542/contact_shadow_issue.zip)

Visual output:
https://github.com/google/filament/assets/37701550/1e126af7-fab1-44ed-abdf-54c93debb1fe

**Reply by romainguy** [reply]
Thanks! I don't get any flickering (macOS, OpenGL, Vulkan, and Metal) :/

**Reply by Nolram12345** [reply]
Then it may be platform related. Might need someone else on Windows & Linux to try it out and attempt to reproduce.
Might be a precision issue, so I do think some kind of slope bias/normal bias for contact shadows could help address that too - if that actually is the issue (but I do think it is seeing as it mainly happens at shallow shadow angles on sloped surfaces).

**Reply by suzp1984** [reply]
which version are you works on?
The maximum allowed `shdow map size` is 1024 at current master branch.

https://github.com/google/filament/blob/2763931b47c86783d0f9c4e375c83d38b6f1ff02/libs/viewer/src/ViewerGui.cpp#L908-L910

Here is my Imgui config:

<img width="377" alt="Screenshot 2024-03-24 at 10 48 25 AM" src="https://github.com/google/filament/assets/2757043/16ac8c80-3127-41c2-b73e-8ec58be04b1b">

I check the git log the 4096 Shadow map size config is long time ago, at least one years ago.

**Reply by Nolram12345** [reply]
I'm using a customised version of the viewer to allow me to tweak some settings that aren't exposed by default such as bias.
However, in regards to resolution, you can just enter a custom number into the shadow map size field, even on the current official version.

I've not changed any rendering code, just the imgui layer to allow me to tweak more settings to see what Filament is capable of and how it performs.

Besides that, this is using the current master.

**Comment by Nolram12345** [comment]
I might escalate this to an issue, as this issue is quite persistent and common when using contact shadows, making them hard to use in production.

**Comment by Nolram12345** [comment]
I discovered that this only happens when TAA is on.

**Comment by Nolram12345** [comment]
Reproduced under Linux & OpenGL using latest release (1.51.3).
Could this perhaps be some driver inconsistency? I'm on NVIDIA drivers 552.12, the latest as of time of writing.
CC @romainguy 

https://github.com/google/filament/assets/37701550/272ecded-2507-4cf6-99de-4c9468658b3b
